### PR TITLE
osd: kill snapdirs

### PR DESCRIPTION
**Pull request by liewegas**
These are gone after luminous.  This rips out all of the compat code from the 
conversion and also simplifies most of hte original code now that head is
always present.

I think SnapSetContext could probably also be removed (use head obc instead),
but haven't tackled that one yet.

Broken into small patches to make each change easier to review.  This will need
a careful look!

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -3605,23 +3605,15 @@ int PrimaryLogPG::trim_object(
...
-    }
-    old_snaps.insert(snapset.clone_snaps[coid.snap].begin(),
-		     snapset.clone_snaps[coid.snap].end());
+  auto citer = snapset.clone_snaps.find(coid.snap);
+  if (citer == snapset.clone_snaps.end()) {
+    osd->clog->error() << "No clone_snaps in snapset " << snapset
+		       << " for object " << coid << "\n";
+    return -ENOENT;
   }
+  set<snapid_t> old_snaps;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: nit, could just put:
```c++
set<snapid_t> old_snaps(citer->second.begin(), citer->second.end());
```

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -3746,14 +3738,10 @@ int PrimaryLogPG::trim_object(
...
     dout(10) << coid << " snaps " << old_snaps << " -> " << new_snaps << dendl;
-    if (legacy) {
-      coi.legacy_snaps = vector<snapid_t>(new_snaps.rbegin(), new_snaps.rend());
-    } else {
-      snapset.clone_snaps[coid.snap] = vector<snapid_t>(new_snaps.rbegin(),
-							new_snaps.rend());
-      // we still do a 'modify' event on this object just to trigger a
-      // snapmapper.update ... :(
-    }
+    snapset.clone_snaps[coid.snap] = vector<snapid_t>(new_snaps.rbegin(),
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: use `emplace()` to save a ctor for temporary vector instance returned by `snapset.clone_snaps[coid.snap]`, like:
```c++
snapset.clone_snaps.emplace(
  coid.snap,
  vector<snapid_t>(new_snaps.rbegin(),new_snaps.rend()));
```

**Comment by tchaikov** [comment]
note to myself, should start from "osd/PrimaryLogPG: trim_object: simplify a bit more for post-snapdir" (inclusive) next time.

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -9956,7 +9955,7 @@ void PrimaryLogPG::add_object_context_to_pg_stat(ObjectContextRef obc, pg_stat_t
   if (oi.is_cache_pinned())
     stat.num_objects_pinned++;
 
-  if (oi.soid.snap && oi.soid.snap != CEPH_NOSNAP && oi.soid.snap != CEPH_SNAPDIR) {
+  if (oi.soid.is_snap()) {
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: yeah, snap == 0 can never happen.

File: src/osd/PrimaryLogPG.cc [hunk 4]
```diff
@@ -11760,51 +11760,6 @@ uint64_t PrimaryLogPG::recover_backfill(
...
-       * snapdir is really just a confusing way to talk about head existing
-       * as a whiteout), but doing that is going to be a somewhat larger
-       * undertaking.
-       *
-       * @see http://tracker.ceph.com/issues/17668
-       */
-      if (!(check.is_head() &&
-	    backfill_info.begin.is_snapdir() &&
-	    check == backfill_info.begin.get_head()))
-	last_backfill_started = check;
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: ah good catch!  fixed

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: y

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: y

**Comment by liewegas** [comment]
Addressed your comments so far, retesting!

File: src/osd/PrimaryLogPG.cc [hunk 5]
```diff
@@ -2070,7 +2070,7 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
   }
 
   ObjectContextRef obc;
-  bool can_create = op->may_write() || op->may_cache();
+  bool can_create = op->may_write();
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: we might want to backport this if this is a bug fix.

File: src/osd/PrimaryLogPG.cc [hunk 6]
```diff
@@ -7303,7 +7303,7 @@ int PrimaryLogPG::prepare_transaction(OpContext *ctx)
   return result;
 }
 
-void PrimaryLogPG::finish_ctx(OpContext *ctx, int log_op_type, bool maintain_ssc)
+void PrimaryLogPG::finish_ctx(OpContext *ctx, int log_op_type)
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: might want to update the commit message, s/can/can't/

File: src/osd/PrimaryLogPG.cc (hunk 5 above)
Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: I think it's pretty harmless... I'd rather not backport it for risk of some unforeseen side-effect.

File: src/osd/PrimaryLogPG.cc [hunk 7]
```diff
@@ -13505,41 +13502,35 @@ void PrimaryLogPG::scrub_snapshot_metadata(
...
-	osd->clog->error() << mode << " " << info.pgid << " " << soid
-			   << " on disk size (" << p->second.size
-			   << ") does not match object info size ("
-			   << oi->size << ") adjusted for ondisk to ("
-			   << pgbackend->be_get_ondisk_size(oi->size)
-			   << ")";
-	soid_error.set_size_mismatch();
-	++scrubber.shallow_errors;
-      }
+    if (pgbackend->be_get_ondisk_size(oi->size) != p->second.size) {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: @liewegas this is not addressed.

File: src/osd/PrimaryLogPG.cc [hunk 8]
```diff
@@ -13626,42 +13291,34 @@ void PrimaryLogPG::scrub_snapshot_metadata(
...
-			   << ") does not match object info size ("
-			   << oi->size << ") adjusted for ondisk to ("
-			   << pgbackend->be_get_ondisk_size(oi->size)
-			   << ")";
-	soid_error.set_size_mismatch();
-	++scrubber.shallow_errors;
-      }
-
-      dout(20) << mode << "  " << soid << " " << oi.get() << dendl;
+    if (pgbackend->be_get_ondisk_size(oi->size) != p->second.size) {
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: need to check `oi` before dereferencing it.

**Comment by tchaikov** [comment]
quite a few tests failed in http://pulpito.ceph.com/kchai-2017-09-18_09:03:46-rados-wip-kefu-testing-2017-09-18-1515-distro-basic-mira/.  will rerun the failed tests in a separated branch.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-09-19_07:36:11-rados-wip-kefu-testing-17579-distro-basic-mira/

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: actually no, we are replacing an existing member, so emplace won't work.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-09-21_21:20:38-rados-wip-sage-testing-2017-09-21-1438-distro-basic-smithi/

**Pull request merged**